http: Cookie extend

zekth opened the pull request:
Add `delCookie` and `setCookie`.

I have a concern about the multiple `Set-Cookie` header. It's common to use BUT `Headers` does not give us the possibility to do this. Do you think any workaround is possible? I think we can keep it like this for the moment and work on multiple `Set-Cookie` headers handling in a future PR.

Also Days / Months format are hardcoded in the module because ATM there is some discussions about embedding or not the ICU in Deno which may break this module. Ref: https://github.com/denoland/deno/issues/1968

ry commented on http/cookie.ts:
```diff
@@ -1,15 +1,105 @@
+  if (rq.headers.has(COOKIE)) {
     const out: Cookie = {};
-    const c = rq.headers.get("Cookie").split(";");
+    const c = rq.headers.get(COOKIE).split(";");
```
I don't think this constant makes thing easier or more maintainable.

zekth replied:
Was because of the `SETCOOKIE` one just uniformise the module. But yep.

ry commented on http/cookie.ts:
```diff
@@ -1,10 +1,97 @@
+    dateString += `${days[opt.Expires.getDay()]}, ${d} ${
+      months[opt.Expires.getUTCMonth()]
+    } ${y} ${h}:${min}:${s} GMT`;
+    out.push(`Expires=${dateString}`);
```
Could you break the date formatting out into a standalone function with tests of its own?

ry commented on http/cookie.ts:
```diff
@@ -19,3 +106,33 @@ export function getCookie(rq: ServerRequest): Cookie {
   return {};
 }
+
+/* Set the cookie header properly in the Response */
```
Use jsdoc style comments.

ry commented on http/cookie.ts:
```diff
@@ -19,3 +106,33 @@ export function getCookie(rq: ServerRequest): Cookie {
+  res.headers.set("Set-Cookie", cookieStringFormat(cookie, opt));
+}
+
+/* Set the cookie header properly in the Response to delete it */
```
Use jsdoc style comments

ry commented:
Regarding the naming of the functions `getCookie`, `setCookie`, and `delCookie`... I wonder if you could research a similar functionality in Go (and Node?) and see what they call these functions? I have the feeling there are better names we could use here.

zekth commented:
About the name consistency for the `delCookie` there is no standard way. Some languages uses `SetCookie` with expiration date way earlier (like it's done here) or `clearCookie` / `rmCookie` / `delCookie` . Depends on the framework. Personnaly after this research i like `clearCookie`

For `setCookie`, example of golang : https://golang.org/pkg/net/http/#SetCookie . But `setCookie` fits our styleguide.

Finally for `getCookie` example in golang : `readCookies` ref: https://golang.org/src/net/http/cookie.go . And concerning node there is no helper for this. For example express add this to its `ServerResponse` class.

ry commented on http/cookie.ts:
```diff
@@ -1,11 +1,103 @@
+
+export type SameSite = "Strict" | "Lax";
+
+export function cookieDateFormat(date: Date): string {
```
Is this a specific date format for cookies? Does it have a name? If it's not just for cookies, consider moving this to //datetime

zekth replied:
I was thinking it was specific but no. It's called `IMF-fixdate`. I'll move it to datetime. Also is `date.toIMF()` good naming for you?

ry replied:
Is there a name for this in Go? Perhaps you can borrow that.

zekth replied:
No, the only mention of it i can find is IMF-fixdate

ry replied:
I think you can call it "RFC1123_GMT"

We don't have the same way of serializing datetime formats as Go. I'd like to slowly move in that direction tho. So I think you can just keep the `toIMF()` where it is now. But please  familiarize yourself with how I think this would be ideally structured:
https://github.com/golang/go/blob/f67f5511ee0f225bcc8943994ba6139eed375e85/src/net/http/server.go#L903-L909
https://github.com/golang/go/blob/f67f5511ee0f225bcc8943994ba6139eed375e85/src/net/http/cookie.go#L200-L204
Consider adding some of the commentary from Go to the jsdoc of "toIMF()"

ry commented on http/cookie.ts:
```diff
@@ -1,11 +1,103 @@
+  Secure?: boolean;
+  HttpOnly?: boolean;
+  SameSite?: SameSite;
+}
```
Use lower-case for the first letter.

It would be nice to have an element like this ([as is done in Go](https://golang.org/pkg/net/http/#Cookie))
```
        unparsed string[] // Raw text of unparsed attribute-value pairs
```

zekth replied:
Refactored and added tests.

ry commented on http/cookie.ts:
```diff
@@ -1,11 +1,103 @@
+  value: string;
+}
+
+export interface CookieOptions {
```
I think this should be called "Cookie" 

https://golang.org/pkg/net/http/#Cookie

zekth replied:
Refactored

ry commented on http/cookie_test.ts:
```diff
@@ -1,25 +1,199 @@
+      maxAge: 2,
+      domain: "deno.land",
+      path: "/",
+      unparsed: ["unparsed=keyvalue", "batman=Bruce"]
```
👍

ry commented:
Go doesn't have a corresponding function to `delCookie`, it just has this:
https://github.com/golang/go/blob/f67f5511ee0f225bcc8943994ba6139eed375e85/src/net/http/cookie.go#L29
Do you think we can also get away with not having `delCookie`?

ry commented on http/cookie.ts:
```diff
@@ -1,14 +1,78 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
```
Add 
```
// Structured similarly to Go's cookie.go
// https://github.com/golang/go/blob/master/src/net/http/cookie.go
```

zekth commented:
@ry about the go Implementation, it's not really common about the manipulation of the `MaxAge`. We can do both i think, because a lot of frameworks use to have `deleteCookie` functions too. I'm ok with both implementation but i think it's more easy to use to have a `delCookie` function.

Just about the `toIMF()` review. You want me to port the code from golang to the `toIMF()` function in a future PR?

ry commented on http/cookie.ts:
```diff
@@ -19,3 +85,52 @@ export function getCookie(rq: ServerRequest): Cookie {
+ *
+ *     delCookie(res,'foo');
+ */
+export function delCookie(res: Response, CookieName: string): void {
```
s/CookieName/cookieName/

ry commented on http/cookie.ts:
```diff
@@ -1,14 +1,80 @@
+ * Parse the cookies of the Server Request
+ * @param rq Server Request
+ */
+export function getCookies(rq: ServerRequest): Cookies {
```
s/rq/req/

ry commented on http/cookie.ts:
```diff
@@ -1,14 +1,80 @@
+
+export type SameSite = "Strict" | "Lax";
+
+function cookieStringFormat(cookie: Cookie): string {
```
s/cookieStringFormat/toString/

zekth commented on http/cookie.ts:
```diff
@@ -19,3 +83,52 @@ export function getCookie(rq: ServerRequest): Cookie {
+    value: "",
+    expires: new Date(0)
+  };
+  res.headers.set("Set-Cookie", cookieStringFormat(c));
```
To delete a cookie you have to set the expire time to earlier. Like golang does is a fallback. Looking at rfc the max age under 0 is invalid

ry replied:
Ok but I think `delCookie` should be a thin wrapper around `setCookie`, because code like this
```
  if (!res.headers) {
    res.headers = new Headers();
  }
```
is repeated.

Sorry for so many nitpicks :)

zekth replied:
no my bad!

ry commented on http/cookie.ts:
```diff
@@ -19,3 +85,49 @@ export function getCookie(rq: ServerRequest): Cookie {
+    value: "",
+    expires: new Date(0)
+  };
+  setCookie(res, c);
```
Final nit: this can be made into one line

motss commented:
@zekth @ry

Correct me if I'm wrong for the following codes:

https://github.com/denoland/deno_std/blob/f1114691038888fc3d8995b64a8028f072569672/http/cookie.ts#L78-L83

In the case of splitting the cookie value with `=` and joining them back with `""`, it will form an incorrect value as compared to the original one. For example, `'PREF=al=en-GB&f1=123; wide=1; SID=123'`.

The expected output should be

```js
{
  PREF: 'al=en-GB&f1=123',
  wide: '1',
  SID: '123'
}
```
instead of 

```js
{
  PREF: alen-GB&f1123,
  wide: '1',
  SID: '123'
}
```

zekth commented:
> @zekth @ry
> 
> Correct me if I'm wrong for the following codes:
> 
> [deno_std/http/cookie.ts](https://github.com/denoland/deno_std/blob/f1114691038888fc3d8995b64a8028f072569672/http/cookie.ts#L78-L83)
> 
> Lines 78 to 83 in [f111469](/denoland/deno_std/commit/f1114691038888fc3d8995b64a8028f072569672)
> 
>  const c = req.headers.get("Cookie").split(";"); 
>  for (const kv of c) { 
>    const cookieVal = kv.split("="); 
>    const key = cookieVal.shift().trim(); 
>    out[key] = cookieVal.join(""); 
>  } 
> In the case of splitting the cookie value with `=` and joining them back with `""`, it will form an incorrect value as compared to the original one. For example, `'PREF=al=en-GB&f1=123; wide=1; SID=123'`.
> 
> The expected output should be
> 
> ```js
> {
>   PREF: 'al=en-GB&f1=123',
>   wide: '1',
>   SID: '123'
> }
> ```
> 
> instead of
> 
> ```js
> {
>   PREF: alen-GB&f1123,
>   wide: '1',
>   SID: '123'
> }
> ```

Right, fair analysis. Should be `out[key] = cookieVal.join(""); `
